File: Data Structures & Algorithms/count-subsequences/submission-4.java
class Solution {
    public int numDistinct(String s, String t) {
        return dfs(s, t, 0, 0);
    }

    private int dfs(String s, String t, int i, int j){
        if(i == s.length() && j == t.length()){
            return 1;
        }

        if(i == s.length()){
            return 0;
        }

        int res = dfs(s, t, i+1, j);
        if(j < t.length() && s.charAt(i) == t.charAt(j)){
            res += dfs(s, t, i+1, j+1);
        }

        return res;
    }
}

File: Data Structures & Algorithms/combinations-of-a-phone-number/submission-4.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        if not digits:
            return []

        numbers = ['', '', 'abc', 'def', 'ghi', 'jkl', 'mno', 'pqrs', 'tuv', 'wxyz']
        res = ['']

        for d in digits:
            temp = []
            for part in res:
                for c in numbers[int(d)]:
                    temp.append(part + c)

            res = temp

        return res